man/zabbix_get.man
.\" Start URL macro. Copied from an-ext.tmac for portability
.de UR
.  ds m1 \\$1\"
.  nh
.  if \\n(mH \{\
.    \" Start diversion in a new environment.
.    do ev URL-div
.    do di URL-div
.  \}
..
.\" End URL macro. Copied from an-ext.tmac for portability
.de UE
.  ie \\n(mH \{\
.    br
.    di
.    ev
.
.    \" Has there been one or more input lines for the link text?
.    ie \\n(dn \{\
.      do HTML-NS "<a href=""\\*(m1"">"
.      \" Yes, strip off final newline of diversion and emit it.
.      do chop URL-div
.      do URL-div
\c
.      do HTML-NS </a>
.    \}
.    el \
.      do HTML-NS "<a href=""\\*(m1"">\\*(m1</a>"
\&\\$*\"
.  \}
.  el \
\\*(la\\*(m1\\*(ra\\$*\"
.
.  hy \\n(HY
..
.TH ZABBIX_GET 1 "2020\-02\-29" Zabbix
.SH NAME
zabbix_get \- Zabbix get utility
.SH SYNOPSIS
.B zabbix_get \-s
.I host\-name\-or\-IP
.RB [ \-p
.IR port\-number ]
.RB [ \-I
.IR IP\-address ]
.BI \-k " item\-key"
.br
.B zabbix_get \-s
.I host\-name\-or\-IP
.RB [ \-p
.IR port\-number ]
.RB [ \-I
.IR IP\-address ]
.B \-\-tls\-connect
.B cert
.B \-\-tls\-ca\-file
.IR CA\-file
.RB [ \-\-tls\-crl\-file
.IR CRL\-file ]
.RB [ \-\-tls\-agent\-cert\-issuer
.IR cert\-issuer ]
.RB [ \-\-tls\-agent\-cert\-subject
.IR cert\-subject ]
.B \-\-tls\-cert\-file
.IR cert\-file
.B \-\-tls\-key\-file
.IR key\-file
.RB [ \-\-tls\-cipher13
.IR cipher\-string ]
.RB [ \-\-tls\-cipher
.IR cipher\-string ]
.BI \-k " item\-key"
.br
.B zabbix_get \-s
.I host\-name\-or\-IP
.RB [ \-p
.IR port\-number ]
.RB [ \-I
.IR IP\-address ]
.B \-\-tls\-connect
.B psk
.B \-\-tls\-psk\-identity
.IR PSK\-identity
.B \-\-tls\-psk\-file
.IR PSK\-file
.RB [ \-\-tls\-cipher13
.IR cipher\-string ]
.RB [ \-\-tls\-cipher
.IR cipher\-string ]
.BI \-k " item\-key"
.br
.B zabbix_get \-h
.br
.B zabbix_get \-V
.SH DESCRIPTION
.B zabbix_get
is a command line utility for getting data from Zabbix agent.
.SH OPTIONS
.IP "\fB\-s\fR, \fB\-\-host\fR \fIhost\-name\-or\-IP\fR"
Specify host name or IP address of a host.
.IP "\fB\-p\fR, \fB\-\-port\fR \fIport\-number\fR"
Specify port number of agent running on the host.
Default is 10050.
.IP "\fB\-I\fR, \fB\-\-source\-address\fR \fIIP\-address\fR"
Specify source IP address.
.IP "\fB\-k\fR, \fB\-\-key\fR \fIitem\-key\fR"
Specify key of item to retrieve value for.
.IP "\fB\-\-tls\-connect\fR \fIvalue\fR"
How to connect to agent. Values:\fR
.SS
.RS 12
.TP 12
.B unencrypted
connect without encryption (default)
.RE
.RS 12
.TP 12
.B psk
connect using TLS and a pre-shared key
.RE
.RS 12
.TP 12
.B cert
connect using TLS and a certificate
.RE
.RE
.IP "\fB\-\-tls\-ca\-file\fR \fICA\-file\fR"
Full pathname of a file containing the top-level CA(s) certificates for peer certificate verification.
.IP "\fB\-\-tls\-crl\-file\fR \fICRL\-file\fR"
Full pathname of a file containing revoked certificates.
.IP "\fB\-\-tls\-agent\-cert\-issuer\fR \fIcert\-issuer\fR"
Allowed agent certificate issuer.
.IP "\fB\-\-tls\-agent\-cert\-subject\fR \fIcert\-subject\fR"
Allowed agent certificate subject.
.IP "\fB\-\-tls\-cert\-file\fR \fIcert\-file\fR"
Full pathname of a file containing the certificate or certificate chain.
.IP "\fB\-\-tls\-key\-file\fR \fIkey\-file\fR"
Full pathname of a file containing the private key.
.IP "\fB\-\-tls\-psk\-identity\fR \fIPSK\-identity\fR"
PSK\-identity string.
.IP "\fB\-\-tls\-psk\-file\fR \fIPSK\-file\fR"
Full pathname of a file containing the pre-shared key.
.IP "\fB\-\-tls\-cipher13\fR \fIcipher\-string\fR"
Cipher string for OpenSSL 1.1.1 or newer for TLS 1.3. Override the default ciphersuite selection criteria. This option is not available if OpenSSL version is less than 1.1.1.
.IP "\fB\-\-tls\-cipher\fR \fIcipher\-string\fR"
GnuTLS priority string (for TLS 1.2 and up) or OpenSSL cipher string (only for TLS 1.2). Override the default ciphersuite selection criteria.
.IP "\fB\-h\fR, \fB\-\-help\fR"
Display this help and exit.
.IP "\fB\-V\fR, \fB\-\-version\fR"
Output version information and exit.
.SH "EXAMPLES"
\fBzabbix_get \-s 127.0.0.1 \-p 10050 \-k "system.cpu.load[all,avg1]"\fR
.br
\fBzabbix_get \-s 127.0.0.1 \-p 10050 \-k "system.cpu.load[all,avg1]" \-\-tls\-connect cert \-\-tls\-ca\-file /home/zabbix/zabbix_ca_file \-\-tls\-agent\-cert\-issuer "CN=Signing CA,OU=IT operations,O=Example Corp,DC=example,DC=com" \-\-tls\-agent\-cert\-subject "CN=server1,OU=IT operations,O=Example Corp,DC=example,DC=com" \-\-tls\-cert\-file /home/zabbix/zabbix_get.crt \-\-tls\-key\-file /home/zabbix/zabbix_get.key
.br
\fBzabbix_get \-s 127.0.0.1 \-p 10050 \-k "system.cpu.load[all,avg1]" \-\-tls\-connect psk \-\-tls\-psk\-identity "PSK ID Zabbix agentd" \-\-tls\-psk\-file /home/zabbix/zabbix_agentd.psk\fR
.SH "SEE ALSO"
.UR https://www.zabbix.com/manuals
Documentation
.UE
.PP
.BR zabbix_agentd (8),
.BR zabbix_proxy (8),
.BR zabbix_sender (8),
.BR zabbix_server (8)
.SH AUTHOR
Alexei Vladishev <alex@zabbix.com>


ChangeLog
Changes for 4.0.19

4.0.19rc1 was released as 4.0.19 without any changes

--------------------------------------------------------------------------------
Changes for 4.0.19rc1

New features:
.......PS. [ZBXNEXT-5649] added name referencing support ('~' suffix) to jsonpath (wiper)
...G...PS. [ZBXNEXT-3604] added support of user configurable ciphers for communication between Zabbix components (Andris)

Bug fixes:
....I..... [ZBX-17173] improved windows agent build scripts (yurii)
A......... [ZBX-17425] fixed template.massupdate API macros update (rlataria)
..F....... [ZBX-17344] fixed PHP 7.4 errors in Latest data->History page for text and log items having an empty filter (Ivo)
.......PS. [ZBX-17330] fixed error reading vmware events with keys over 2^31 (MVekslers)
..F....... [ZBX-17288] fixed undefined indexes in Administration->Regular expressions->Test tab (averza, Ivo)
..F....... [ZBX-17350] fixed displaying of multiple value types in Latest data->History page (Ivo)
..FG...... [ZBX-16802] added mtime-reread/mtime-noreread options parameter to log* keys (atumilovics)
..F....... [ZBX-17390] fixed validation of the maximum selected period (Sasha)
..F....... [ZBX-17185] fixed non-disappearing messages of failed login attempts (Ivo)
........S. [ZBX-17322] fixed deadlock between server and frontend when updating multiple hosts (vso)
..F.I..... [ZBX-9057] fixed duplicate entries being written into in profiles table (talbergs)
..F....... [ZBX-17159] fixed overlay dialogue instantiation (talbergs)
.......PS. [ZBX-17108] improved process utilization calculation (mabele)
..F....... [ZBX-17269] fixed undefined indexes in history page (Ivo)
........S. [ZBX-17342] dump memory statistics and items with most values when value cache is fully used (vso)
.......PS. [ZBX-16926] fix INT_MAX error when compiling on SunOS/Joyent/Solaris (arimdjonoks)

--------------------------------------------------------------------------------
Changes for 4.0.18

4.0.18rc1 was released as 4.0.18 without any changes

--------------------------------------------------------------------------------
Changes for 4.0.18rc1

New features:
..F....PS. [ZBXNEXT-5603,ZBXNEXT-5626] added database character set check (mabele, rlataria)
.......PS. [ZBXNEXT-5690] added support for libssh (akozlovs)
....I..... [ZBXNEXT-5706] replaced xml2-config to pkg-config when building with libxml2 (vso)

Bug fixes:
.......P.. [ZBX-17213] fixed Zabbix proxy not to execute remote commands on Zabbix agent with timeout and not to wait for execution result (vso)
........S. [ZBX-17201] fixed snmp trapper processes exceeding 1000% (arimdjonoks)
A.F....... [ZBX-17220] fixed several PHP 7.4 compatibility issues (miks)
........S. [ZBX-16827] fixed applications being unlinked from items when both items and applications are not discovered anymore (vso)
........S. [ZBX-17035] internal events are not created if there is no enabled internal actions (atumilovics)
........S. [ZBX-16965] fixed link/unlink template operation execution order (atumilovics)

--------------------------------------------------------------------------------
Changes for 4.0.17

4.0.17rc1 was released as 4.0.17 without any changes

--------------------------------------------------------------------------------
Changes for 4.0.17rc1

New features:
...G...... [ZBXNEXT-5490] added new key vfs.fs.get to collect mounted filesystems information and relevant metrics into json (akozlovs)
........S. [ZBXNEXT-5239] added housekeeping of unused/deleted items values in value cache (vso)
A......... [ZBX-3783] added strict validation of input parameters in script.get() method (Sasha)

Bug fixes:
A.......S. [ZBX-10240] fixed maintenance time period update every field validation (atumilovics)
...G...... [ZBX-17164] fixed file change time in vfs.file.time on Windows (atumilovics)
...G...... [ZBX-17036] fixed memory leak and wrong type cast; thanks to Yudai Hashimoto for the patch (mabele)
A.......S. [ZBX-16811] fixed SQL error during prototype removal by adding select for update locks (atumilovics)
.....J.... [ZBX-17002] added bulk gathering of attributes for Zabbix Java gateway when using attribute discovery (viktors)
...G...PS. [ZBX-16960,ZBX-16961,ZBX-16964] fixed time of check - time of use issues reported by coverity (atumilovics)
..F....... [ZBX-16902] fixed providing notifications for devices without audio support (talbergs)
........S. [ZBX-17127] fixed crash on jsonpath function processing (atumilovics)
........S. [ZBX-17080] fixed deadlock on maintenance table when using Oracle database (vso)
A.F....... [ZBX-17078] fixed web scenario step allowing to use 0 timeout (mabele, talbergs)
.......PS. [ZBX-16907] fixed icmppingloss counting all after first 400 ping responses as lost (mabele)
..F....... [ZBX-16325] fixed httpstepid validation when its value exceeds int32 (miks, vasilijs)
........S. [ZBX-16997] fixed duplicate entry errors on 32-bit architecture during item application discovery (akozlovs)
..F....... [ZBX-17102] fixed scroll bar visibility in svg graph widget configuration form (gcalenko)
A......... [ZBX-17049] fixed non well formed numeric value encountered in maintenance (gcalenko)
..F....... [ZBX-16481] fixed spelling issues in the code (miks)
.......... [ZBX-1977] reverted fix for axis labels calculation (rlataria)
..F....... [ZBX-16894] fixed dynamic rows losing old input at form error (talbergs)
........S. [ZBX-16966] fixed not operator in correlation function expression (atumilovics)

--------------------------------------------------------------------------------
Changes for 4.0.16

4.0.16rc1 was released as 4.0.16 without any changes

--------------------------------------------------------------------------------
Changes for 4.0.16rc1

New features:
A.F....... [ZBXNEXT-5271] implemented 'delete missing' option for imported template linkages (miks)

Bug fixes:
..F....... [ZBX-17101] fixed url validation before output (miks)

src/zabbix_get/zabbix_get.c
/*
** Zabbix
** Copyright (C) 2001-2020 Zabbix SIA
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
**/

#include "common.h"

#include "threads.h"
#include "comms.h"
#include "cfg.h"
#include "log.h"
#include "zbxgetopt.h"
#include "../libs/zbxcrypto/tls.h"

#ifndef _WINDOWS
#	include "zbxnix.h"
#endif

const char	*progname = NULL;
const char	title_message[] = "zabbix_get";
const char	syslog_app_name[] = "zabbix_get";
const char	*usage_message[] = {
	"-s host-name-or-IP", "[-p port-number]", "[-I IP-address]", "-k item-key", NULL,
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	"-s host-name-or-IP", "[-p port-number]", "[-I IP-address]", "--tls-connect cert", "--tls-ca-file CA-file",
	"[--tls-crl-file CRL-file]", "[--tls-agent-cert-issuer cert-issuer]", "[--tls-agent-cert-subject cert-subject]",
	"--tls-cert-file cert-file", "--tls-key-file key-file",
#if defined(HAVE_OPENSSL)
	"[--tls-cipher13 cipher-string]",
#endif
#if defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	"[--tls-cipher cipher-string]",
#endif
	"-k item-key", NULL,
	"-s host-name-or-IP", "[-p port-number]", "[-I IP-address]", "--tls-connect psk",
	"--tls-psk-identity PSK-identity", "--tls-psk-file PSK-file",
#if defined(HAVE_OPENSSL)
	"[--tls-cipher13 cipher-string]",
#endif
#if defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	"[--tls-cipher cipher-string]",
#endif
	"-k item-key", NULL,
#endif
	"-h", NULL,
	"-V", NULL,
	NULL	/* end of text */
};

unsigned char	program_type	= ZBX_PROGRAM_TYPE_GET;

const char	*help_message[] = {
	"Get data from Zabbix agent.",
	"",
	"General options:",
	"  -s --host host-name-or-IP  Specify host name or IP address of a host",
	"  -p --port port-number      Specify port number of agent running on the host",
	"                             (default: " ZBX_DEFAULT_AGENT_PORT_STR ")",
	"  -I --source-address IP-address   Specify source IP address",
	"",
	"  -k --key item-key          Specify key of the item to retrieve value for",
	"",
	"  -h --help                  Display this help message",
	"  -V --version               Display version number",
	"",
	"TLS connection options:",
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	"  --tls-connect value        How to connect to agent. Values:",
	"                               unencrypted - connect without encryption",
	"                                             (default)",
	"                               psk         - connect using TLS and a pre-shared",
	"                                             key",
	"                               cert        - connect using TLS and a",
	"                                             certificate",
	"",
	"  --tls-ca-file CA-file      Full pathname of a file containing the top-level",
	"                             CA(s) certificates for peer certificate",
	"                             verification",
	"",
	"  --tls-crl-file CRL-file    Full pathname of a file containing revoked",
	"                             certificates",
	"",
	"  --tls-agent-cert-issuer cert-issuer   Allowed agent certificate issuer",
	"",
	"  --tls-agent-cert-subject cert-subject   Allowed agent certificate subject",
	"",
	"  --tls-cert-file cert-file  Full pathname of a file containing the certificate",
	"                             or certificate chain",
	"",
	"  --tls-key-file key-file    Full pathname of a file containing the private key",
	"",
	"  --tls-psk-identity PSK-identity   Unique, case sensitive string used to",
	"                             identify the pre-shared key",
	"",
	"  --tls-psk-file PSK-file    Full pathname of a file containing the pre-shared",
	"                             key",
#if defined(HAVE_OPENSSL)
	"",
	"  --tls-cipher13             Cipher string for OpenSSL 1.1.1 or newer for",
	"                             TLS 1.3. Override the default ciphersuite",
	"                             selection criteria. This option is not available",
	"                             if OpenSSL version is less than 1.1.1",
#endif
#if defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	"",
	"  --tls-cipher               GnuTLS priority string (for TLS 1.2 and up) or",
	"                             OpenSSL cipher string (only for TLS 1.2).",
	"                             Override the default ciphersuite selection",
	"                             criteria",
#endif
#else
	"  Not available. This 'zabbix_get' was compiled without TLS support",
#endif
	"",
	"Example(s):",
	"  zabbix_get -s 127.0.0.1 -p " ZBX_DEFAULT_AGENT_PORT_STR " -k \"system.cpu.load[all,avg1]\"",
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	"",
	"  zabbix_get -s 127.0.0.1 -p " ZBX_DEFAULT_AGENT_PORT_STR " -k \"system.cpu.load[all,avg1]\" \\",
	"    --tls-connect cert --tls-ca-file /home/zabbix/zabbix_ca_file \\",
	"    --tls-agent-cert-issuer \\",
	"    \"CN=Signing CA,OU=IT operations,O=Example Corp,DC=example,DC=com\" \\",
	"    --tls-agent-cert-subject \\",
	"    \"CN=server1,OU=IT operations,O=Example Corp,DC=example,DC=com\" \\",
	"    --tls-cert-file /home/zabbix/zabbix_get.crt \\",
	"    --tls-key-file /home/zabbix/zabbix_get.key",
	"",
	"  zabbix_get -s 127.0.0.1 -p " ZBX_DEFAULT_AGENT_PORT_STR " -k \"system.cpu.load[all,avg1]\" \\",
	"    --tls-connect psk --tls-psk-identity \"PSK ID Zabbix agentd\" \\",
	"    --tls-psk-file /home/zabbix/zabbix_agentd.psk",
#endif
	NULL	/* end of text */
};

/* TLS parameters */
unsigned int	configured_tls_connect_mode = ZBX_TCP_SEC_UNENCRYPTED;
unsigned int	configured_tls_accept_modes = ZBX_TCP_SEC_UNENCRYPTED;	/* not used in zabbix_get, just for linking */
									/* with tls.c */
char	*CONFIG_TLS_CONNECT		= NULL;
char	*CONFIG_TLS_ACCEPT		= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CA_FILE		= NULL;
char	*CONFIG_TLS_CRL_FILE		= NULL;
char	*CONFIG_TLS_SERVER_CERT_ISSUER	= NULL;
char	*CONFIG_TLS_SERVER_CERT_SUBJECT	= NULL;
char	*CONFIG_TLS_CERT_FILE		= NULL;
char	*CONFIG_TLS_KEY_FILE		= NULL;
char	*CONFIG_TLS_PSK_IDENTITY	= NULL;
char	*CONFIG_TLS_PSK_FILE		= NULL;

char	*CONFIG_TLS_CIPHER_CERT13	= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CIPHER_CERT		= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CIPHER_PSK13	= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CIPHER_PSK		= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CIPHER_ALL13	= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CIPHER_ALL		= NULL;	/* not used in zabbix_get, just for linking with tls.c */
char	*CONFIG_TLS_CIPHER_CMD13	= NULL;	/* parameter '--tls-cipher13' from zabbix_get command line */
char	*CONFIG_TLS_CIPHER_CMD		= NULL;	/* parameter '--tls-cipher' from zabbix_get command line */

int	CONFIG_PASSIVE_FORKS		= 0;	/* not used in zabbix_get, just for linking with tls.c */
int	CONFIG_ACTIVE_FORKS		= 0;	/* not used in zabbix_get, just for linking with tls.c */

/* COMMAND LINE OPTIONS */

/* long options */
struct zbx_option	longopts[] =
{
	{"host",			1,	NULL,	's'},
	{"port",			1,	NULL,	'p'},
	{"key",				1,	NULL,	'k'},
	{"source-address",		1,	NULL,	'I'},
	{"help",			0,	NULL,	'h'},
	{"version",			0,	NULL,	'V'},
	{"tls-connect",			1,	NULL,	'1'},
	{"tls-ca-file",			1,	NULL,	'2'},
	{"tls-crl-file",		1,	NULL,	'3'},
	{"tls-agent-cert-issuer",	1,	NULL,	'4'},
	{"tls-agent-cert-subject",	1,	NULL,	'5'},
	{"tls-cert-file",		1,	NULL,	'6'},
	{"tls-key-file",		1,	NULL,	'7'},
	{"tls-psk-identity",		1,	NULL,	'8'},
	{"tls-psk-file",		1,	NULL,	'9'},
	{"tls-cipher13",		1,	NULL,	'A'},
	{"tls-cipher",			1,	NULL,	'B'},
	{NULL}
};

/* short options */
static char	shortopts[] = "s:p:k:I:hV";

/* end of COMMAND LINE OPTIONS */

#if !defined(_WINDOWS)

/******************************************************************************
 *                                                                            *
 * Function: get_signal_handler                                               *
 *                                                                            *
 * Purpose: process signals                                                   *
 *                                                                            *
 * Parameters: sig - signal ID                                                *
 *                                                                            *
 * Return value:                                                              *
 *                                                                            *
 * Comments:                                                                  *
 *                                                                            *
 ******************************************************************************/
/******************************************************************************
 * *
 *这段代码的主要目的是定义一个信号处理函数`get_signal_handler`，用于处理进程接收到的信号。当进程接收到SIGPIPE（客户端关闭连接信号）或SIGALRM（超时信号）时，该函数会根据不同的信号类型进行相应的处理，如释放TLS资源或输出错误信息。最后，进程以失败退出。
 ******************************************************************************/
/* 定义一个信号处理函数，用于处理进程接收到的信号
 * 参数：sig 接收到的信号
 */
static void get_signal_handler(int sig)
{
	// 判断接收到的信号是否为SIGPIPE，即客户端关闭连接信号
	if (SIGPIPE == sig) {
		/* 当对端因为访问限制关闭连接时，会发送SIGPIPE信号 */
		return;
	}

	// 判断接收到的信号是否为SIGALRM，即超时信号
	if (SIGALRM == sig) {
		zbx_error("Timeout while executing operation");
	}

	// 判断是否使用了TLS加密
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	if (ZBX_TCP_SEC_UNENCRYPTED != configured_tls_connect_mode) {
		/* 在信号处理函数中释放TLS资源 */
		zbx_tls_free_on_signal();
	}
#endif

/******************************************************************************
 * *
 *整个代码块的主要目的是实现一个名为`get_value`的函数，该函数用于通过TCP连接向指定的服务器发送关键字并接收返回数据。在这个过程中，根据配置的TLS连接模式设置相应的参数，并在连接成功后处理接收到的数据。最后，将处理后的数据输出到控制台或记录错误信息。函数返回连接状态（成功或失败）。
 ******************************************************************************/
static int	get_value(const char *source_ip, const char *host, unsigned short port, const char *key)
{
	/* 定义一个zbx_socket_t类型的结构体变量s，用于存储套接字信息 */
	zbx_socket_t	s;
	/* 定义一个int类型的变量ret，用于存储函数返回值 */
	int		ret;
	/* 定义一个ssize_t类型的变量bytes_received，初始值为-1 */
	ssize_t		bytes_received = -1;
	/* 定义两个char类型的指针变量tls_arg1和tls_arg2，用于存储TLS相关参数 */
	char		*tls_arg1, *tls_arg2;

	/* 判断configured_tls_connect_mode的值，根据不同值设置tls_arg1和tls_arg2的值 */
	switch (configured_tls_connect_mode)
	{
		case ZBX_TCP_SEC_UNENCRYPTED:
			tls_arg1 = NULL;
			tls_arg2 = NULL;
			break;
		/* 如果系统支持POLARSSL、GNUTLS或OPENSSL，则根据configured_tls_connect_mode的值设置tls_arg1和tls_arg2的值 */
		case ZBX_TCP_SEC_TLS_CERT:
			tls_arg1 = CONFIG_TLS_SERVER_CERT_ISSUER;
			tls_arg2 = CONFIG_TLS_SERVER_CERT_SUBJECT;
			break;
		case ZBX_TCP_SEC_TLS_PSK:
			tls_arg1 = CONFIG_TLS_PSK_IDENTITY;
			tls_arg2 = NULL;	/* zbx_tls_connect()会自动查找PSK */
			break;
		default:
			/* 如果不支持当前的configured_tls_connect_mode，则执行THIS_SHOULD_NEVER_HAPPEN宏，返回FAIL */
			THIS_SHOULD_NEVER_HAPPEN;
			return FAIL;
	}

	/* 使用zbx_tcp_connect()函数尝试连接目标服务器，如果成功则继续执行后续操作 */
	if (SUCCEED == (ret = zbx_tcp_connect(&s, source_ip, host, port, GET_SENDER_TIMEOUT,
			configured_tls_connect_mode, tls_arg1, tls_arg2)))
	{
		/* 如果发送关键字key到服务器成功，则继续接收服务器返回的数据 */
		if (SUCCEED == (ret = zbx_tcp_send(&s, key)))
		{
			/* 接收服务器返回的数据，并保存在bytes_received中，如果接收到的数据长度大于0，则继续处理 */
			if (0 < (bytes_received = zbx_tcp_recv_ext(&s, 0)))
			{
				/* 判断接收到的数据是否为ZBX_NOTSUPPORTED，如果是则处理后续逻辑 */
				if (0 == strcmp(s.buffer, ZBX_NOTSUPPORTED) && sizeof(ZBX_NOTSUPPORTED) < s.read_bytes)
				{
					/* 去除字符串尾部的换行符和空格，然后输出到控制台 */
					zbx_rtrim(s.buffer + sizeof(ZBX_NOTSUPPORTED), "\r\
");
					printf("%s: %s\
", s.buffer, s.buffer + sizeof(ZBX_NOTSUPPORTED));
				}
				else
				{
					/* 去除字符串尾部的换行符和空格，然后输出到控制台 */
					zbx_rtrim(s.buffer, "\r\
");
					printf("%s\
", s.buffer);
				}
			}
			else
			{
				/* 如果接收到的数据长度为0，则判断是否需要检查访问限制配置 */
				if (0 == bytes_received)
					zbx_error("Check access restrictions in Zabbix agent configuration");
				/* 如果接收到的数据长度不为0，但连接失败，则记录错误信息并返回FAIL */
				ret = FAIL;
			}
		}

		/* 关闭套接字连接 */
		zbx_tcp_close(&s);

		/* 如果连接失败且接收到的数据长度不为0，则输出错误信息并提示检查访问限制配置 */
		if (SUCCEED != ret && 0 != bytes_received)
		{
/******************************************************************************
 * *
 *这个代码块的主要目的是解析命令行参数，并根据解析结果初始化相关变量。其中，程序接收主机名、密钥、源IP地址等参数，并根据这些参数调用`get_value`函数获取值。同时，程序还处理了TLS相关的选项，并在需要时初始化TLS连接。最后，程序释放内存并返回成功或失败的状态。
 ******************************************************************************/
int main(int argc, char **argv)
{
	// 定义变量，用于存储主机名、密钥、源IP地址等
	int i, ret = SUCCEED;
	char *host = NULL, *key = NULL, *source_ip = NULL, ch;
	unsigned short opt_count[256] = {0}, port = ZBX_DEFAULT_AGENT_PORT;

	// 检查是否定义了TLS相关参数，如果有，进行初始化
#if defined(_WINDOWS)
	char *error = NULL;
#endif

	// 解析命令行参数
	while ((char)EOF != (ch = (char)zbx_getopt_long(argc, argv, shortopts, longopts, NULL)))
	{
		// 统计每个选项出现的次数
		opt_count[(unsigned char)ch]++;

		switch (ch)
		{
			// 处理各种选项
			case 'k':
				if (NULL == key)
					key = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'p':
				port = (unsigned short)atoi(zbx_optarg);
				break;
			case 's':
				if (NULL == host)
					host = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'I':
				if (NULL == source_ip)
					source_ip = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'h':
				help();
				exit(EXIT_SUCCESS);
				break;
			case 'V':
				version();
				exit(EXIT_SUCCESS);
				break;
			// TLS相关选项
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
			case '1':
				CONFIG_TLS_CONNECT = zbx_strdup(CONFIG_TLS_CONNECT, zbx_optarg);
				break;
			case '2':
				CONFIG_TLS_CA_FILE = zbx_strdup(CONFIG_TLS_CA_FILE, zbx_optarg);
				break;
			case '3':
				CONFIG_TLS_CRL_FILE = zbx_strdup(CONFIG_TLS_CRL_FILE, zbx_optarg);
				break;
			case '4':
				CONFIG_TLS_SERVER_CERT_ISSUER = zbx_strdup(CONFIG_TLS_SERVER_CERT_ISSUER, zbx_optarg);
				break;
			case '5':
				CONFIG_TLS_SERVER_CERT_SUBJECT = zbx_strdup(CONFIG_TLS_SERVER_CERT_SUBJECT, zbx_optarg);
				break;
			case '6':
				CONFIG_TLS_CERT_FILE = zbx_strdup(CONFIG_TLS_CERT_FILE, zbx_optarg);
				break;
			case '7':
				CONFIG_TLS_KEY_FILE = zbx_strdup(CONFIG_TLS_KEY_FILE, zbx_optarg);
				break;
			case '8':
				CONFIG_TLS_PSK_IDENTITY = zbx_strdup(CONFIG_TLS_PSK_IDENTITY, zbx_optarg);
				break;
			case '9':
				CONFIG_TLS_PSK_FILE = zbx_strdup(CONFIG_TLS_PSK_FILE, zbx_optarg);
				break;
			case 'A':
				break;
			case 'B':
				break;
			// 默认处理未知选项
			default:
				usage();
				exit(EXIT_FAILURE);
				break;
		}
	}

	// 检查选项是否存在重复
	for (i = 0; NULL != longopts[i].name; i++)
	{
		ch = longopts[i].val;

		if (1 < opt_count[(unsigned char)ch])
		{
			if (NULL == strchr(shortopts, ch))
				zbx_error("option \"--%s\" specified multiple times", longopts[i].name);
			else
				zbx_error("option \"-%c\" or \"--%s\" specified multiple times", ch, longopts[i].name);

			ret = FAIL;
		}
	}

	// 如果有错误，退出程序
	if (FAIL == ret)
	{
		printf("Try '%s --help' for more information.\
", progname);
		goto out;
	}

	// 处理非选项参数
	if (argc > zbx_optind)
	{
		for (i = zbx_optind; i < argc; i++)
			zbx_error("invalid parameter \"%s\"", argv[i]);

		ret = FAIL;
	}

	// 调用get_value函数获取值
	ret = get_value(source_ip, host, port, key);

out:
	// 释放内存
	zbx_free(host);
	zbx_free(key);
	zbx_free(source_ip);

	// 清理TLS相关资源
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	if (ZBX_TCP_SEC_UNENCRYPTED != configured_tls_connect_mode)
	{
		zbx_tls_free();
#if defined(_WINDOWS)
		zbx_tls_library_deinit();
#endif
	}
#endif

	// 处理Windows下的套接字初始化
#if defined(_WINDOWS)
	while (0 == WSACleanup())
		;
#endif

	// 返回程序状态
	return SUCCEED == ret ? EXIT_SUCCESS : EXIT_FAILURE;
}

			NULL != CONFIG_TLS_PSK_IDENTITY || NULL != CONFIG_TLS_PSK_FILE ||
			NULL != CONFIG_TLS_CIPHER_CMD13 || NULL != CONFIG_TLS_CIPHER_CMD)
	{
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
		zbx_tls_validate_config();

		if (ZBX_TCP_SEC_UNENCRYPTED != configured_tls_connect_mode)
		{
#if defined(_WINDOWS)
			zbx_tls_init_parent();
#endif
			zbx_tls_init_child();
		}
#endif
	}
#if !defined(_WINDOWS)
	signal(SIGINT, get_signal_handler);
	signal(SIGQUIT, get_signal_handler);
	signal(SIGTERM, get_signal_handler);
	signal(SIGHUP, get_signal_handler);
	signal(SIGALRM, get_signal_handler);
	signal(SIGPIPE, get_signal_handler);
#endif
	ret = get_value(source_ip, host, port, key);
out:
	zbx_free(host);
	zbx_free(key);
	zbx_free(source_ip);
#if defined(HAVE_POLARSSL) || defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
	if (ZBX_TCP_SEC_UNENCRYPTED != configured_tls_connect_mode)
	{
		zbx_tls_free();
#if defined(_WINDOWS)
		zbx_tls_library_deinit();
#endif
	}
#endif
#if defined(_WINDOWS)
	while (0 == WSACleanup())
		;
#endif

	return SUCCEED == ret ? EXIT_SUCCESS : EXIT_FAILURE;
}
